### PR TITLE
Add SONAME/SOVERSION to generated library 

### LINKED ISSUE #229: CMake-built shared library is missing SONAME / library versioning
I tried to go through open and closed issues and couldn't find it, but I it could well be that this is a duplicate.

Anyway, the exported CMake build via `-DLIBINT2_BUILD_SHARED_AND_STATIC_LIBS=ON` just builds `libint2.so`, no library versioning.

2.7.1:

```
libint2-2.7.1# objdump -x obj-x86_64-linux-gnu/libint2.so | grep SONAME   
  SONAME               libint2.so
```

2.6.0:

```
objdump -x usr/lib/x86_64-linux-gnu/libint2.so.2.0.3 | grep SONAME
  SONAME               libint2.so.2
```

I browsed through `CMakeLists.txt` and couldn't find anything about library versioning either, am I missing something?

cc @loriab though I have the nagging feeling we discussed this already...

What I think needs changing is this:

```
--- ./export/cmake/CMakeLists.txt.export.orig   2022-01-01 12:43:08.071905766 +0100
+++ ./export/cmake/CMakeLists.txt.export        2022-01-01 12:47:00.621475434 +0100
@@ -194,7 +194,7 @@
 # shared and static libraries built from the same object files
 if (LIBINT2_BUILD_SHARED_AND_STATIC_LIBS)
   add_library(libint2 SHARED $<TARGET_OBJECTS:libint2_obj>)
-  set_target_properties(libint2 PROPERTIES LIBRARY_OUTPUT_NAME int2)
+  set_target_properties(libint2 PROPERTIES LIBRARY_OUTPUT_NAME int2 VERSION 2.0.4 SOVERSION 2)
   add_library(libint2-static STATIC $<TARGET_OBJECTS:libint2_obj>)
   set_target_properties(libint2-static PROPERTIES ARCHIVE_OUTPUT_NAME int2)
   target_include_directories(libint2-static INTERFACE
```

This leads to:

```
ibint2-2.7.1# ls -l obj-x86_64-linux-gnu/libint2.so*
lrwxrwxrwx 1 mba mba       12 Jan  1 11:58 obj-x86_64-linux-gnu/libint2.so -> libint2.so.2
lrwxrwxrwx 1 mba mba       16 Jan  1 11:58 obj-x86_64-linux-gnu/libint2.so.2 -> libint2.so.2.0.4
-rwxr-xr-x 1 mba mba 71004840 Jan  1 11:58 obj-x86_64-linux-gnu/libint2.so.2.0.4
```
and

```
libint2-2.7.1# objdump -x obj-x86_64-linux-gnu/libint2.so | grep SONAME   
  SONAME               libint2.so.2
```

However, at this point, it is not yet clear to me what `VERSION` actually should be, i.e. whether 2.7.1 breaks API with 2.6.0 (then `SOVERSION` should be bumped to 3) or at least adds to API (then `SOVERSION` should stay at 2, but `VERSION` should be bumped to 2.1.0).

### PR DESCRIPTION
**Pull request by mbanck**
Fixes #229

### CODE REVIEW

**Comment by loriab** [comment]
Might be more robust to pull from https://github.com/evaleev/libint/blob/master/configure.ac#L3 akin to https://gitlab.com/libxc/libxc/-/blob/master/CMakeLists.txt#L125-135 and https://gitlab.com/libxc/libxc/-/blob/master/CMakeLists.txt#L473-474 .

I don't think the `VERSION` here should be the current:revision:age https://cmake.org/pipermail/cmake-developers/2016-December/029695.html . Would just the `SOVERSION` be ok?

**Comment by mbanck** [comment]
Well `current:revision:age` is terribly convoluted and maps to some `x.y.z` number scheme (see the libxc Gitlab link; the `[2:3:0]` in libint's `configure.ac` maps to `2.0.3`). It is my understanding that `VERSION` is supposed to be that `x.y.z` (see e.g. https://mail.kde.org/pipermail/kde-buildsystem/2008-April/004543.html) and `SOVERSION` the `SONAME` number encoded in the library, and also the single-number API-level `.x` ending which gets symlinked.

I didn't realize the top-level `configure.ac` still has the `libint_so_version`; as this one did not get bumped for 2.7.1, I guess `VERSION` should be 2.0.3 then, not 2.0.4 as in this PR.

Also note that libxc does not seem to use `${PROJECT_NAME}_SOVERSION` at all in its build system (other than to output it during configuring) and it is setting `SOVERSION` to `${PROJECT_NAME}_SOMAJOR` in https://gitlab.com/libxc/libxc/-/blob/master/CMakeLists.txt#L474

**Comment by mbanck** [comment]
One more comment: As `configure.ac` is (to the best of my knowledge) not being pulled into the exported generated library tarball, it cannot be used by CMake similar to what LibXC does, that information just isn;t available I think.

Also, I guess the future is all-CMake, so this is a stop-gap measure. I would agree that it is pretty brittle if we would have to use it for a long time and always remember to bump the library version in both places, but this sounds like a one-off (or two-off) think, famous last words...

**Comment by evaleev** [comment]
@mbanck thanks ... I tried to understand what `VERSION` and `SOVERSION` are supposed to be (CMake's dox are no bueno) but I am still not sure about the best-practices:
- It seems to me that since `VERSION` is for human consumption it is easiest to set it to the project version or to whatever the package maintainers (i.e. you) need it to be. I don't understand why it is set to `2.0.4` then ...
- I gave up on API/ABI version tracking in this and other projects because it is far too easy to change the ABI (plus it is unrealistic to test ABI, etc.). Libint library's API/ABI is very stable but I cannot guarantee that it has only changed twice (is that what `SOVERSION 2` means?) ... also: the `SOVERSION` implies library configuration (max AM, etc.) so if those change in the future `SOVERSION` would have to change?

**Comment by StefanBruens** [comment]
>     * I gave up on API/ABI version tracking in this and other projects because it is far too easy to change the ABI (plus it is unrealistic to test ABI, etc.). Libint library's API/ABI is very stable but I cannot guarantee that it has only changed twice (is that what `SOVERSION 2` means?) ... also: the `SOVERSION` implies library configuration (max AM, etc.) so if those change in the future `SOVERSION` would have to change?

SOVERSION is part of the SONAME. It should be bumped at least every time the ABI changes in a not-backwards compatible way. In case symbols are not versioned (e.g. using a linker version script, as is the case here), the SOVERSION should also be bumped every time the ABI is changed in a backwards compatible way (e.g. when adding new functions).

The library configuration is an orhogonal aspect. IMHO each configuration should have a different SONAME, and multiple libraries could be installed in parallel. According to https://github.com/evaleev/libint/issues/190, there are 20 different configurations, though only 5 seem to be actually used. E.g. psi4 could link to libint2_gss.

File: export/cmake/CMakeLists.txt.export [hunk 1]
```diff
@@ -194,7 +194,7 @@ endif()
 # shared and static libraries built from the same object files
 if (LIBINT2_BUILD_SHARED_AND_STATIC_LIBS)
   add_library(libint2 SHARED $<TARGET_OBJECTS:libint2_obj>)
-  set_target_properties(libint2 PROPERTIES LIBRARY_OUTPUT_NAME int2)
+  set_target_properties(libint2 PROPERTIES LIBRARY_OUTPUT_NAME int2 VERSION 2.0.4 SOVERSION 2)
```
Comment (evaleev) [inline export/cmake/CMakeLists.txt.export:197]: would `${LIBINT_VERSION}` would for `VERSION`?

**Comment by evaleev** [comment]
good starter-level explanation of `SONAME` and `SOVERSION`: https://crascit.com/wp-content/uploads/2019/09/Deep-CMake-For-Library-Authors-Craig-Scott-CppCon-2019.pdf

File: export/cmake/CMakeLists.txt.export (hunk 1 above)
Comment (StefanBruens) [reply export/cmake/CMakeLists.txt.export:197]: As the API/ABI are not strictly controlled for backwards compatibility, the SOVERSION should be the same as the VERSION (better safe than sorry ...). This is the CMake default if SOVERSION is omitted.

```
set_target_properties(libint2 PROPERTIES
    LIBRARY_OUTPUT_NAME int2
    VERSION ${LIBINT_VERSION}
)
```

Comment (evaleev) [reply export/cmake/CMakeLists.txt.export:197]: Thanks. I think `SOVERSION 2` is appropriate, the ABI of the library has been stable for many years.

**Comment by loriab** [comment]
I don't think this PR thoroughly added SOVERSION (see below for mac/linux with BUILD_SHARED_LIBS=ON). I haven't fully investigated, but the PR may have just caught the build-both-static-and-shared target.

```
2023-12-19T20:24:34.6214061Z -rwxr-xr-x  2 conda conda  4790576 Dec 19 20:23 libint2.so
2023-12-19T20:24:34.6214631Z lrwxrwxrwx  1 conda conda       15 Dec 19 20:24 libitm.so -> libitm.so.1.0.0
2023-12-19T20:24:34.6215059Z lrwxrwxrwx  1 conda conda       15 Dec 19 20:24 libitm.so.1 -> libitm.so.1.0.0
2023-12-19T20:24:34.6215442Z -rwxrwxr-x  1 conda conda   773744 Nov 12 01:46 libitm.so.1.0.0
```
```
2023-12-19T20:33:15.3850370Z -rwxr-xr-x    2 runner  staff    4923120 Dec 19 20:30 libint2.dylib
2023-12-19T20:33:15.3851270Z -rwxrwxr-x    4 runner  staff      75584 Aug 15 11:12 libk5crypto.3.1.dylib
2023-12-19T20:33:15.3902810Z lrwxr-xr-x    1 runner  staff         21 Dec 19 20:33 libk5crypto.3.dylib -> libk5crypto.3.1.dylib
2023-12-19T20:33:15.3952710Z lrwxr-xr-x    1 runner  staff         21 Dec 19 20:33 libk5crypto.dylib -> libk5crypto.3.1.dylib
```

**Pull request merged**